### ISSUE TITLE
callback registered with unit manager not called

### ISSUE THREAD

**Issue by vivek-bala** [radical-cybertools/radical.pilot#1358]
I have a callback that I registered with the unit manager which does not seem to be called during CU state transitions. I get the following warning:

```
2017-05-31 10:35:13,839: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000001: StagingInput)
2017-05-31 10:35:13,839: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000001: AgentStagingInputPending)
2017-05-31 10:35:13,840: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000001: AgentStagingInput)
2017-05-31 10:35:13,840: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000001: AllocatingPending)
2017-05-31 10:35:13,840: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000001: Allocating)
2017-05-31 10:35:13,840: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000001: ExecutingPending)
2017-05-31 10:35:13,841: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000001: Executing)
2017-05-31 10:35:13,841: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000001: AgentStagingOutputPending)
2017-05-31 10:35:13,842: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000001: AgentStagingOutput)
2017-05-31 10:35:13,842: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000001: PendingOutputStaging)
2017-05-31 10:35:13,843: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000000: AgentStagingInput)
2017-05-31 10:35:13,843: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000000: AllocatingPending)
2017-05-31 10:35:13,843: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000000: Allocating)
2017-05-31 10:35:13,844: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000000: ExecutingPending)
2017-05-31 10:35:13,844: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000000: Executing)
2017-05-31 10:35:13,845: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000000: AgentStagingOutputPending)
2017-05-31 10:35:13,845: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000000: AgentStagingOutput)
2017-05-31 10:35:13,845: radical.pilot       : MainProcess                     : Thread-4       : WARNING : skip cb for incomple unit (unit.000000: PendingOutputStaging)
```

Upon checking the sandbox, the units do complete successfully. But the callbacks don't execute and hence some of the other threads stall forever.

Link to the callback code: https://github.com/vivek-bala/radical.entk-0.6/blob/feature/add_rp/src/radical/entk/execman/task_manager.py#L97-L130
Line where the callback is registered: https://github.com/vivek-bala/radical.entk-0.6/blob/feature/add_rp/src/radical/entk/execman/task_manager.py#L155

I am using the released version of RP and the target resource is localhost. 

Am I missing something? Thanks

**Comment by vivek-bala** [comment]
Also, just wanted to mention that there are no callbacks in the example scripts. So this error will be no caught while testing by users and/or jenkins.

**Comment by andre-merzky** [comment]
Hmm, interesting. But that code path has completely changed in devel, and I am thus somewhat hesitant to start debugging it in master.  Can you give devel a try with that script, please, or is there a compelling reason to fix this for the old code?

**Comment by vivek-bala** [comment]
Not really, no particular reason to use Master apart from stability. I tried devel with the first example but I get a different error:

```
Traceback (most recent call last):
  File "00_getting_started.py", line 9, in <module>
    import radical.pilot as rp
  File "/home/vivek/Research/ves/test_rp_entk/local/lib/python2.7/site-packages/radical/pilot/__init__.py", line 23, in <module>
    from .session                   import Session
  File "/home/vivek/Research/ves/test_rp_entk/local/lib/python2.7/site-packages/radical/pilot/session.py", line 29, in <module>
    from . import utils         as rpu
  File "/home/vivek/Research/ves/test_rp_entk/local/lib/python2.7/site-packages/radical/pilot/utils/__init__.py", line 15, in <module>
    from .queue        import *
  File "/home/vivek/Research/ves/test_rp_entk/local/lib/python2.7/site-packages/radical/pilot/utils/queue.py", line 102, in <module>
    class Queue(ru.Process):
AttributeError: 'module' object has no attribute 'Process'
```

**Comment by andre-merzky** [comment]
You will need to update RU to devel, too, in this case - and also RS, FWIW.

**Comment by vivek-bala** [comment]
After installing the devel stack, I get the following error while submitting the pilot:

bootstrap_1.out
```
# -------------------------------------------------------------------
purge install source at radical.pilot-split-74-g10acb0d-devel/
verify python viability: /home/vivek/radical.pilot.sandbox/ve_localhost/bin/python ... ok
verify module viability: saga            ... ok
verify module viability: radical.utils   ... ok
verify module viability: pymongo         ... ok
verify module viability: hostlist        ... ok
verify module viability: netifaces       ... ok
verify module viability: setproctitle    ... ok
verify module viability: ntplib          ... ok
verify module viability: msgpack         ...Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module named msgpack
 failed
python installation cannot load module msgpack - abort
```

Full log: https://gist.github.com/vivek-bala/4455c89b42ef843e376492631dd26a9e

**Comment by marksantcroos** [comment]
Try removing `/home/vivek/radical.pilot.sandbox/ve_localhost` and then running again.

**Comment by vivek-bala** [comment]
Thanks Mark. That takes it part that issue. What is the verbose env variable to be used with devel? RADICAL_PILOT_VERBOSE=INFO does not seem to work.

**Comment by andre-merzky** [comment]
Ah, bugger - this deserves a new ticket I'm afraid - now opened as #1363.  Thanks Mark for jumping in: cleaning out the VE will indeed work around the issue.

`RADICAL_PILOT_VERBOSE` should actually work.  What changed in devel though is that many subcomponents now write to their individual log files, in a subdirectory named with the session id.  If that does not work, please open a separate issue.  Thanks!

**Comment by andre-merzky** [comment]
#1363 is not addressed via #1367, and thus the original problem should be solved.  Waiting on feedback on the `VERBOSE` settings.

**Comment by vivek-bala** [comment]
The verbose setting does indeed create the log files in a folder. However, I still can't get the callback to be triggered upon state change. Andre, can we schedule a call to discuss this further.

**Comment by andre-merzky** [comment]
This was caused by creating a umgr object, then calling fork(), then using the umgr object in the child process.  The umgr will loose its message channels during the fork, and basically be deaf and blind.  Surprisingly, the DB connection seemed to survive, so no error was triggered during unit submission.  Oh well.

I think we should not go the extra mile to try to inhibit this usage - `atfork()` handlers are difficult to handle on a per-object basis, anything else would be ugly.  Thus closing the ticket, in the hope that nobody else tries this :P

**Comment by vivek-bala** [comment]
Thanks Andre for helping debug!
